design/StableFunctions.md

```diff
@@ -0,0 +1,126 @@
+# Stable Functions, Stable Objects, and Stable Classes
+
+Functions, objects, and classes can also be persisted with enhanced orthogonal persistence, i.e. they been used in Motoko's stable variables and persistent actor variables.
+
+Note: This is not supported with classical persistence.
+
+With this, the only non-persistent (non-stable) values remain to be lambdas, async handles (aka continuations, futures), and local async function references.
+
+## Example
+
+With this, the following program can now be used for orthogonal persistence. For this example, you need a [specific base library branch](https://github.com/dfinity/motoko-base/tree/luc%2Fstable-functions), which contains slight modifications for supporting stable functions.
+
+```
+import HashMap "base/HashMap";
+import Hash "base/Hash";
+import Nat "base/Nat";
+import Iter "base/Iter";
+
+persistent actor {
+    let hashMap = HashMap.HashMap<Nat, Text>(10, Nat.equal, Hash.hash); // finally persistent!
+
+    public func populate(): async() {
+        hashMap.put(1, "A");
+        hashMap.put(2, "B");
+        hashMap.put(3, "C");
+    };
+
+    public func list(): async [(Nat, Text)] {
+        Iter.toArray<(Nat, Text)>(hashMap.entries());
+    };
+}
+```
+
+## Stable Functions and Stable Scopes
+
+A stable function is a named non-async local function in a stable scope, only closing over variables of a stable type.
+
+A stable scope is:
+* the main actor,
+* an actor class,
+* a module imported with a unique identifier from a stable scope,
+* a named non-async function in a stable scope,
+* a class in a stable scope, or,
+* a named object in a stable scope.
+
+Generic type parameters of stable functions and stable classes are bounded to stable types.
+
+A stable function is also a stable type.
+
+Syntactically, function types are prefixed by `stable` to denote a stable function, e.g.
+`stable X -> Y`. Stable functions implicitly have a corresponding stable reference type.
+
+A stable function type is a sub-type of a flexible function type with type-compatible signature, i.e. `stable X' -> Y <: X -> Y'` for `X' <: X` and `Y' :< Y`.
```

Comment:
Generic functions require some more care. You probably want the type parameters to match (exactly), with equivalent sorts and bounds.

Reply:
True. A specification of generics would be needed (as currently implemented):
* A stable function with type parameters cannot be promoted to a flexible functions. This is because stable functions have implicitly stable type parameters, while flexible functions have flexible type parameters.
* Bounds need to match. This should already be the case with classical implementation.